### ISSUE TITLE
New configuration syntax introduced in `10.1` minor release is not backwards compatible

### ISSUE THREAD

**Issue by ghostwriter** [sebastianbergmann/phpunit#5335]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 10.1.0
| PHP version         | >=8.1
| Installation Method | Composer

#### Summary

In version `10.1.0` of PHPUnit, a new configuration syntax has been introduced that is not backwards compatible with previous versions. This change was introduced in pull request #5294.

And the PHPUnit configuration migration feature is not working properly. (I've not had time to pinpoint the exact issue)

We're seeing, most if not all, PHPUnit-related CI actions failing in @laminas, @mezzio, and our personal repositories.

https://github.com/laminas/laminas-code/actions/runs/4716480790/jobs/8364248270
https://github.com/laminas/laminas-migration/actions/runs/4721162509/jobs/8374119988
https://github.com/mezzio/mezzio-problem-details/actions/runs/4716453001/jobs/8364196408
https://github.com/mezzio/mezzio-authorization-rbac/actions/runs/4716159258/jobs/8363633788
https://github.com/ghostwriter/event-dispatcher/actions/runs/4722678948/jobs/8377614377
https://github.com/sebastianbergmann/phpunit/issues/5294#issuecomment-1506904338
https://github.com/sebastianbergmann/phpunit/issues/5332#issue-1669970642

#### Current behavior

Tests fail due to **various** syntax errors in the configuration file.

```
phpunit.xml.dist:13: element include: Schemas validity error : Element 'include': This element is not expected.
phpunit.xml.dist fails to validate
```

```
There was 1 PHPUnit test runner warning:

1) Test results may not be as expected because the XML configuration file did not pass validation:

  Line 14:
  - Element 'include': This element is not expected.
```

Tools like `infection/infection` that use PHPUnit under the hood fail:
```
The file "/tmp/infection/phpunitConfiguration.initial.infection.xml" does not pass the XSD schema validation.
[Fatal] Start tag expected, '<' not found in https://schema.phpunit.de/10.0/phpunit.xsd (line 1, col 1)
Error: r] Failed to parse the XML resource 'https://schema.phpunit.de/10.0/phpunit.xsd'.
```

Migrating the configuration file fails (not always):

```
Migration failed: "./phpunit.xml" is not a valid PHPUnit XML configuration file that can be migrated
```

```
Migration failed: "mezzio-helpers/phpunit.xml.dist" is not a valid PHPUnit XML configuration file that can be migrated
```

#### How to reproduce

1. Update to the latest minor release (`10.1.x`) of PHPUnit
2. Run PHPUnit
3. Notice one of the various error messages

#### Expected behavior

1. Update to the latest minor release (`10.1.x`) of PHPUnit
2. Attempt to run tests using a configuration file in the old syntax
3. Notice a message to migrate the configuration file.
4. Tests should run using the old configuration syntax without issue.

OR

1. Update to the latest minor release (`10.1.x`) of PHPUnit
2. Attempt to run tests using a configuration file in the old syntax
3. Notice a message to migrate the configuration file.
4. Run `phpunit --migrate-configuration`
5. Attempt to run tests using a configuration file in the new syntax
6. Tests should run using the new configuration syntax without issue.

#### Proposed Solution

 - Revert the changes
 - Update the PHPUnit documentation to include the new configuration syntax
 - Deprecate the old syntax in `10.1`
 - Remove the old syntax in `11`.

**Comment by sebastianbergmann** [comment]
> Tools like `infection/infection` that use PHPUnit under the hood fail:

You need to update Infection: https://github.com/infection/infection/issues/1840

**Comment by sebastianbergmann** [comment]
> In version `10.1.0` of PHPUnit, a new configuration syntax has been introduced that is not backwards compatible with previous versions.

That is not true: a `phpunit.xml` configuration file that validates against the PHPUnit 10.0 `phpunit.xsd` continues to work with PHPUnit 10.1, but you get a deprecation warning.

The old configuration syntax is deprecated in PHPUnit 10.1 and [will be removed in PHPUnit 11](https://github.com/sebastianbergmann/phpunit/issues/5329). The documentation has been [updated](https://github.com/sebastianbergmann/phpunit-documentation-english/commit/b990d4e171af1a7a3f6c2a7af1b068e2297b56d4) to explain the new configuration syntax and no longer shows the old syntax.

The CI pipelines you mention fail because they use `xmllint --schema vendor/phpunit/phpunit/phpunit.xsd phpunit.xml.dist` to validate `phpunit.xml` configuration files against the `phpunit.xsd` schema file of the installed PHPUnit version. That will obviously fail.

When PHPUnit loads an XML configuration file it tries to validate it first against the schema of the current PHPUnit version. When it validates: great! When it does not then it tried to validate against previous versions' schemas.

**Comment by sebastianbergmann** [comment]
I am sorry to say this, but as it is, this ticket is not actionable. If you have reproducing test cases for `phpunit --migrate` not working correctly then please open (a) separate ticket(s) for that.

**Comment by ghostwriter** [comment]
@sebastianbergmann, Thank you for taking the time to review this issue and providing a detailed response.

> The old configuration syntax is deprecated in PHPUnit 10.1

Deprecating  `10.0` configuration syntax should not cause `PHPUnit 10.1` to fail validation (only show the deprecation message).

I believe both `10.0` and `10.1` syntax should be backward compatible (no validation failure, only deprecation notice).

As for a reproducible test case for `phpunit --migrate`, similar to #5324

Thanks again.

**Comment by sebastianbergmann** [comment]
> Deprecating `10.0` configuration syntax should not cause `PHPUnit 10.1` to fail validation (only show the deprecation message).

And here I disagree: While PHPUnit 10.1 "fails" to validate an XML configuration file that is valid for PHPUnit 10.0, it loads it just like PHPUnit 10.0 would have.

> As for a reproducible test case for `phpunit --migrate`, similar to #5324

I am sorry, but "similar to some other bug" is not actionable. If you provide a minimal, self-contained, reproducing test case that fails right now with PHPUnit 10.1.1 (unlike #5324 which was already fixed in PHPUnit 10.1.0) then I will be more than happy into fixing `--migrate` issues.
